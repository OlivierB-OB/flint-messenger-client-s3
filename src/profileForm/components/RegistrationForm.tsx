import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import React from 'react';
import { connect } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { updateProfileForm } from '../actions/updateProfileForm';
import { IProfileFormFields, IProfileFormStatus } from '../types';
import { IdentitySection } from './IdentitySection';
import { CredentialsSection } from './CredentialsSection';
import { Alert } from '../../layout/components/Alert';
import { makeResetRegistrationForm } from '../actions/makeResetRegistrationForm';
import { makeSubmitRegistrationForm } from '../actions/makeSubmitRegistrationForm';

export interface IRegistrationDisplayFormProps {
  status: IProfileFormStatus;
  fields: IProfileFormFields;
  update<T extends keyof IProfileFormFields>(field: T, value: IProfileFormFields[T]['value']): void;
  saveProfile(): void;
}

export function RegistrationFormDisplay({ status, fields, update, saveProfile }: IRegistrationDisplayFormProps) {
  const { email, firstName, lastName, password, confirmation } = fields;
  return (
    <Container>
      <Box style={{ margin: '2rem 0' }}>
        <Alert status={status} />
      </Box>
      <Box style={{ margin: '2rem 0' }}>
        <Grid container justify="space-evenly" alignItems="flex-start">
          <Grid item xs={4}>
            <IdentitySection
              allowEmailEdition={true}
              email={email}
              firstName={firstName}
              lastName={lastName}
              update={update}
            />
          </Grid>
          <Grid item xs={4}>
            <CredentialsSection password={password} confirmation={confirmation} update={update} />
          </Grid>
        </Grid>
      </Box>
      <Box style={{ margin: '2rem 0' }}>
        <Grid container justify="flex-end">
          <Grid item xs={2}>
            <Button variant="contained" color="primary" fullWidth={true} onClick={saveProfile}>
              Register
            </Button>
          </Grid>
        </Grid>
      </Box>
    </Container>
  );
}

const mapStateToProps = ({ profileForm }: IAppState) => ({
  status: profileForm.status,
  fields: profileForm.fields,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  update: <T extends keyof IProfileFormFields>(field: T, value: IProfileFormFields[T]['value']) =>
    dispatch(updateProfileForm(field, value)),
  resetForm: () => dispatch(makeResetRegistrationForm()),
  saveProfile: () => dispatch(makeSubmitRegistrationForm()),
});

export const RegistrationForm = connect(mapStateToProps, mapDispatchToProps)(RegistrationFormDisplay);
