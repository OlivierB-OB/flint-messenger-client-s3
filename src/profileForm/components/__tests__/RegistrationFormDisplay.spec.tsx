import React from 'react';
import { shallow } from 'enzyme';
import Button from '@material-ui/core/Button';
import { RegistrationFormDisplay } from '../RegistrationForm';

describe('RegistrationFormDisplay', () => {
  it('should display the profile form', async () => {
    const props: any = {
      status: 'the status',
      fields: {
        email: 'email field',
        firstName: 'firstName field',
        lastName: 'lastName field',
        password: 'password field',
        confirmation: 'confirmation field',
      },
      update: 'update',
      saveProfile: 'saveProfile',
    };
    const component = shallow(<RegistrationFormDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should allow saving', async () => {
    const props: any = {
      status: 'ready',
      fields: {},
      saveProfile: jest.fn(),
    };
    const component = shallow(<RegistrationFormDisplay {...props} />);
    expect(component.find(Button).at(0).prop('type')).toEqual('submit');
    component.find('form').simulate('submit');
    expect(props.saveProfile).toHaveBeenCalled();
  });
});
