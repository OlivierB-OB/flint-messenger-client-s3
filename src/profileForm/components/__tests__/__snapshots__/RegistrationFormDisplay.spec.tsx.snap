// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationFormDisplay should display the profile form 1`] = `
<WithStyles(ForwardRef(Container))>
  <Styled(MuiBox)
    style={
      Object {
        "margin": "2rem 0",
      }
    }
  >
    <Alert
      status="the status"
    />
  </Styled(MuiBox)>
  <form
    onSubmit="saveProfile"
  >
    <Styled(MuiBox)
      style={
        Object {
          "margin": "2rem 0",
        }
      }
    >
      <WithStyles(ForwardRef(Grid))
        alignItems="flex-start"
        container={true}
        justify="space-evenly"
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={4}
        >
          <IdentitySection
            allowEmailEdition={true}
            email="email field"
            firstName="firstName field"
            lastName="lastName field"
            update="update"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={4}
        >
          <CredentialsSection
            confirmation="confirmation field"
            password="password field"
            update="update"
          />
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </Styled(MuiBox)>
    <Styled(MuiBox)
      style={
        Object {
          "margin": "2rem 0",
        }
      }
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        justify="flex-end"
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={2}
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            fullWidth={true}
            type="submit"
            variant="contained"
          >
            Register
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </Styled(MuiBox)>
  </form>
</WithStyles(ForwardRef(Container))>
`;
